Simplenote/src/main/java/com/automattic/simplenote/utils/AccountVerificationWatcher.java

```diff
@@ -18,57 +18,63 @@
  * have (and probably was) updated from another device.
  */
 public class AccountVerificationWatcher implements Bucket.OnNetworkChangeListener<Account> {
-    private enum Status {
-        UNKNOWN,
+    public enum Status {
+        SENT_EMAIL,
         UNVERIFIED,
         VERIFIED,
     }
 
+    public interface VerificationStateListener {
+        void onUpdate(Status status);
+    }
+
     private final Simplenote simplenote;
-    private Status currentStatus = Status.UNKNOWN;
+    private final List<VerificationStateListener> listeners = new ArrayList<>();
 
     public AccountVerificationWatcher(Simplenote simplenote) {
         this.simplenote = simplenote;
     }
 
-    public void clearStatus() {
-        currentStatus = Status.UNKNOWN;
+    public void addVerificationStateListener(VerificationStateListener listener) {
+        listeners.add(listener);
     }
 
-    @Override
-    public void onNetworkChange(Bucket<Account> bucket, Bucket.ChangeType type, String key) {
-        // If we already know the answer, we avoid computing it again
-        if (currentStatus != Status.UNKNOWN) {
-            return;
+    private void notifyListeners(Status status) {
+        for (VerificationStateListener listener : listeners) {
+            listener.onUpdate(status);
         }
+    }
+
+    @Override
+    public void onNetworkChange(final Bucket<Account> bucket, Bucket.ChangeType type, String key) {
+        // If the account is removed, the status is changed to UNVERIFIED immediately since the account will not be
+        // available
+       if (type == Bucket.ChangeType.REMOVE) {
```

Comment:
I am about 60% confident that this change isn't saying "the bucket has been removed" but rather "the entity with key `key` has been removed." it's possible that someone might add or remove another entity in this bucket. currently inside the `account` bucket we use the entity whose entity id is `email-verification` to store this information. that would make this check look a little different…

```java
if (type == Bucket.ChangeType.REMOVE && key == KEY_EMAIL_VERIFICATION) {
    notifyListeners(Status.UNVERIFIED);
    return;
}
```

Reply:
@dmsnell Question(❓ ): currently in every update of `onNetworkChange` I see that `key` is `null` (mostly for `RESET` and `INDEX` events). Maybe because those are general events instead of events on particular keys such as `REMOVE`?

Reply:
that's right. `key` will be a string value, the `entity_id` of an entity, for those network events which refer to an entity. these would be `REMOVE`, `INSERT`, `MODIFY`. `RESET` is kind of a made-up event that's not coming from the network but is part of the Android Simperium library, used to reset bucket storage because of the aforementioned issues where these aren't being destroyed on signet.